src/keywrap/keyprovider.rs

```diff
@@ -0,0 +1,572 @@
+                };
+            }
+        }
+        let key_wrap_results = match protocol_output.key_wrap_results {
+            Some(x) => x,
+            None => return Err(anyhow!("key-provider: {:?} on {:?} operation results are empty", self.provider, OP_KEY_WRAP.to_string()))
+        };
+
+        Ok(key_wrap_results.annotation)
+    }
+
+    /// UnwrapKey calls appropriate binary-executable or grpc/ttrpc server for unwrapping the session key based on the protocol given in annotation for recipients and gets decrypted optsData,
+    /// which describe the symmetric key used for decrypting the layer
+    fn unwrap_keys(&self, dc_config: &DecryptConfig, json_string: &[u8]) -> Result<Vec<u8>> {
+        let mut protocol_output = KeyProviderKeyWrapProtocolOutput { key_wrap_results: None, key_unwrap_results: None };
+        let key_unwrap_params = KeyUnwrapParams { dc: Some(dc_config.clone()), annotation: Vec::from(json_string) };
+        let input = KeyProviderKeyWrapProtocolInput {
+            op: OP_KEY_UNWRAP.to_string(),
+            key_wrap_params: None,
+            key_unwrap_params: Option::from(key_unwrap_params),
+        };
+        let serialized_input = match serde_json::to_vec(&input) {
+            Ok(x) => x,
+            Err(_) => return Err(anyhow!("Error while serializing key provider input parameters on {:?} operation", OP_KEY_UNWRAP.to_string()))
+        };
+
+        if self.attrs.cmd.as_ref().is_some() {
+            protocol_output = match get_provider_command_output(serialized_input, &self.attrs.cmd, self.runner.as_ref().unwrap()) {
+                Ok(v) => v,
+                Err(e) => return Err(anyhow!("Error while key provider {:?} operation, from binary executable provider, error: {:?}", OP_KEY_UNWRAP.to_string(), e))
+            };
+        } else if self.attrs.grpc.as_ref().is_some() {
+            let rt = Runtime::new().unwrap().block_on(get_provider_grpc_output(serialized_input, self.attrs.grpc.as_ref().unwrap(), OP_KEY_UNWRAP.to_string()));
+            protocol_output = match rt {
+                Ok(v) => v,
+                Err(e) => return Err(anyhow!("Error while key provider {:?} operation, from grpc provider error, error: {:?}", OP_KEY_UNWRAP.to_string(), e))
+            };
+        }
+        let key_unwrap_results = match protocol_output.key_unwrap_results {
+            Some(x) => x,
+            None => return Err(anyhow!("keyprovider: {:?} on {:?} operation results are empty",self.provider,OP_KEY_UNWRAP.to_string()))
+        };
+
+        Ok(key_unwrap_results.opts_data)
+    }
+
+
+    fn annotation_id(&self) -> String {
+        format!("{}{}", "org.opencontainers.image.enc.keys.provider.", self.provider)
+    }
+
+    fn no_possible_keys(&self, _dc_param: &HashMap<String, Vec<Vec<u8>>, RandomState>) -> bool {
+        unimplemented!()
+    }
+}
+
+async fn get_provider_grpc_output(input: Vec<u8>, conn: &str, operation: String) -> Result<KeyProviderKeyWrapProtocolOutput> {
```

Comment:
Here as well, you want to follow a light Builder pattern and have the following construct:

```Rust
impl KeyProviderKeyWrapProtocolOutput {
        fn from_grpc(input: Vec<u8>, conn: &str) -> Result<Self> {
            [...]
        }

       fn from_command(input: Vec<u8>, cmd: &Option<Command>) -> Result<Self> {
           [...]
       }
}
```

instead of using ad-hoc functions that are a little more difficult to read.

Reply:
yes done